collectors/CMPCollector.js

```diff
@@ -328,8 +414,28 @@ class CMPCollector extends BaseCollector {
      * @returns {Promise<CMPResult[]>}
      */
     async getData() {
+        const patterns = await this.scanPatterns();
```

Comment:
Please correct me if I'm missing something, here's what I believe happens:

- when `getSiteData()` calls `collector.getData()`, it waits for as long as the collector needs.
- Inside `CMPCollector.getData()`, after calling `scanPatterns()`, we wait for `waitForFinish()`, which in turn has 2s timeout on every step. This is enough for CPM, and `getData()` will not block forever
- if `getData()` takes too long, `getSiteData()` will be interrupted by `maxTotalTimeMs`. If this happens, crawlerConductor will retry
- If retries don't help, the whole site is considered a failure, and it will not produce a file with a result
- The number of captured page crawls seems [large enough](https://grafana.duckduckgo.com/goto/TNOnnZnVz?orgId=1). As I understand, this means everything is ok, right?

Reply:
on second thought, those coverage crawls are only using `cmps,screenshots` as collectors, so it might still be a problem for TR crawls. But we could still see it by the number of resulting files

Reply:
Gotcha, thanks for breaking this down, AFAIK you are correct.

As you know we have four phases in the crawler
- init and collector setup phase
- page load phase (networkidle0 or maxLoadTimeMs)
- js execution phase (extraExecutionTimeMs)
- data collection phase and cleanup

I'm still worried that:

(a) crawls will take much longer because now we are handling the prompt not in the js execution phase (meant for stuff like CPM), but data collection phase (meant to be quick)
(b) we are loosing the big benefit that the CMP collector was bringing - getting information about trackers that are loaded after cookie prompt was handled i.e. the prompt is handled, but it's possible that it happens after we alrady called getData on other collectors or there will be not enough time for other collectors to register the changes.

Can we possibly scan for patterns earlier and try to opt out ASAP? Should we introduce another hook so that collectors are pinged when page load phase finishes? Or should this collector wait for page load itself to start scanning and opt out?

Reply:
I see, those are good points. I'll see what I can do